File: Go/1650.LowestCommonAncestorBinaryTreeIII/Solution.go
/* Two pointers: Time:O(h), Space:O(1)
 * 1. Use the same logic of LeetCode#160, since the the distance of reaching LCA on second time is the same
 */

package main

import (
    "fmt"
)

 type Node struct {
    Val int
    Left *Node
    Right *Node
    Parent *Node
 }

func lowestCommonAncestor(p *Node, q *Node) *Node {
    ptr0 := p
    ptr1 := q
    for ptr0 != ptr1 {
        if(ptr0.Parent == nil){
            ptr0 = q
        }else{
            ptr0 = ptr0.Parent
        }

        if(ptr1.Parent == nil){
            ptr1 = p
        }else{
            ptr1 = ptr1.Parent
        }
    }
    return ptr0
}


func main() {
    root := &Node{5, nil, nil, nil}
    p := &Node{2, nil, nil, nil}
    q := &Node{-3, nil, nil, nil}
    root.Left = p
    p.Parent = root
    root.Right = q
    q.Parent = root
    fmt.Println("LCA of", p.Val ,"and", q.Val, "is:", lowestCommonAncestor(p, q).Val)
}

File: Java/1650.LowestCommonAncestorBinaryTreeIII/Solution.java
/* Two pointers: Time:O(h), Space:O(1)
 * 1. Use the same logic of LeetCode#160, since the the distance of reaching LCA on second time is the same
 */

import java.util.*; // Stack

/* Definition for binary tree */
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node parent;
    public Node(int v){val = v;}
}

public class Solution {
    public Node lowestCommonAncestor(Node p, Node q) {
        Node ptr0 = p;
        Node ptr1 = q;
        while(ptr0 != ptr1) {
            ptr0 =(ptr0.parent == null)? q: ptr0.parent;
            ptr1 =(ptr1.parent == null)? p: ptr1.parent;
        }
        return ptr0;
    }
  
    public static void main(String[] args){
        Solution sol = new Solution();
        
        /* Generate a input tree
         *     5
         *    / \
         *   2   -5
         */
        Node root = new Node(5);
        root.left = new Node(2);
        root.left.parent = root;
        root.right = new Node(-5);
        root.right.parent = root;

        Node p = root.left;
        Node q = root.right;
        System.out.println("LCA of " + p.val + " and " + q.val + " is:" + sol.lowestCommonAncestor(p, q).val );
    }
}
